include/lsst/afw/typehandling/Storable.h

```diff
@@ -0,0 +1,158 @@
+#include "lsst/afw/table/io/Persistable.h"
+
+namespace lsst {
+namespace afw {
+namespace typehandling {
+
+/**
+ * Exception thrown by Storable operations for unimplemented operations.
+ *
+ * As with all RuntimeError, callers should assume that this exception may be thrown at any time.
+ */
+LSST_EXCEPTION_TYPE(UnsupportedOperationException, pex::exceptions::RuntimeError,
+                    lsst::afw::typehandling::UnsupportedOperationException)
+
+/**
+ * Interface supporting iteration over heterogenous containers.
+ *
+ * Many operations defined by Storable are optional, and may throw UnsupportedOperationException if they are
+ * not defined.
+ *
+ * @see StorableHelper
+ *
+ * @note All Storables are equality-comparable through operator==(Storable const&, Storable const&). This may
+ * cause inconsistent behavior when Storable is used as a mixin in a class hierarchy with an existing equality
+ * operator. Developers should take care to ensure the result is the same no matter which operator is called;
+ * disambiguating calls may require adding an explicit overload for the derived class.
+ */
+class Storable : public table::io::Persistable {
+public:
+    virtual ~Storable() noexcept = 0;
+
+    /**
+     * Create a new object that is a copy of this one (optional operation).
+     *
+     * @throws UnsupportedOperationException Thrown if this object is not cloneable.
+     */
+    virtual std::unique_ptr<Storable> clone() const {
+        throw LSST_EXCEPT(UnsupportedOperationException, "Cloning is not supported.");
+    }
+
+    /**
+     * Create a string representation of this object (optional operation).
+     *
+     * @throws UnsupportedOperationException Thrown if this object does not have a string representation.
+     */
+    virtual std::string toString() const {
+        throw LSST_EXCEPT(UnsupportedOperationException, "No string representation available.");
+    }
+
+    /**
+     * Return a hash of this object (optional operation).
+     *
+     * @throws UnsupportedOperationException Thrown if this object is not hashable.
+     *
+     * @note Subclass authors are responsible for any associated specializations of std::hash.
+     */
+    virtual std::size_t hash_value() const {
+        throw LSST_EXCEPT(UnsupportedOperationException, "Hashes are not supported.");
+    }
+
+protected:
+    friend bool operator==(Storable const& lhs, Storable const& rhs) noexcept;
+
+    /**
+     * Compare this object to another Storable.
+     *
+     * @return This implementation always returns `false`.
+     *
+     * @warning If this operation is defined, then subclasses must be comparable to any type of Storable
+     * (although cross-class comparisons should usually return `false`). If cross-class comparisons are valid,
+     * implementers should take care that they are symmetric.
+     */
+    virtual bool equals(Storable const& other) const noexcept { return false; }
+};
+
+inline Storable::~Storable() {}
```

Comment:
I think so? I defined one for all my classes because my IDE complained about a non-virtual destructor if I didn't, but I'm pretty sure that it simply wasn't applying the rules for implicitly defined destructors correctly.